File: 021_merge_two_sorted_lists.py
# Merge two sorted linked lists and return it as a new list.
# The new list should be made by splicing together the nodes of the first two lists.
"""
class Solution(object):
    def mergeTwoLists(self, a, b):
        if a and b:
            if a.val > b.val:
                a, b = b, a
            a.next = self.mergeTwoLists(a.next, b)
        return a or b
"""

# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

    def __repr__(self):
        if self:
            return "{} -> {}".format(self.val, repr(self.next))

class Solution(object):
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        dummy = current = ListNode(0)

        while l1 and l2:
            if l1.val <= l2.val:
                current.next, current, l1 = l1, l1, l1.next
            else:
                current.next, current, l2 = l2, l2, l2.next
        # the rest elements will append to the list
        if not l1:
            current.next = l2
        else:
            current.next = l1
        #current.next = l1 or l2

        return dummy.next

if __name__ == '__main__':
    l1 = ListNode(2)
    l1.next = ListNode(5)
    l1.next.next = ListNode(7)
    l2 = ListNode(1)
    l2.next = ListNode(3)
    l2.next.next = ListNode(4)
    print(Solution().mergeTwoLists(l1, l2))

File: C++/021_merge_two_sorted_lists.cpp
// Merge two sorted linked lists and return it as a new list.
// The new list should be made by splicing together the nodes of the first two lists.

#include <iostream>

// Definition for single-linked list
struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if (!l1) return l2;
        if (!l2) return l1;

        ListNode* current = new ListNode(0);
        ListNode* dummy = current;

        while (l1 && l2) {
            if (l1->val <= l2->val) {
                current->next = l1;
                current       = l1;
                l1            = l1->next;
            }
            else {
                current->next = l2;
                current       = l2;
                l2            = l2->next;
            }
        }
        if (!l1) {
            current->next = l2;
        }
        if (!l2) {
            current->next = l1;
        }

        return dummy->next;
    }
};


int main() {
    Solution Test;
    ListNode *l1 = new ListNode(1);
    l1->next = new ListNode(2);
    l1->next->next = new ListNode(5);
    ListNode *l2 = new ListNode(1);
    l2->next = new ListNode(3);
    l2->next->next = new ListNode(4);

    ListNode *result = Test.mergeTwoLists(l1, l2);

    while (result) {
        std::cout << result->val << std::endl;
        result = result->next;
    }

    // std::cout << "the merged sorted list is: " << Test.mergeTwoLists(l1, l2) << std::endl;

    return 0;
}
